src/tools/tool-options.h

```diff
@@ -24,98 +24,109 @@
+                         errorOnFeatureMismatch = false;
+                       })
+        .addFeature("sign-ext", "sign extension operations",
+                    FeatureSet::SignExt)
+        .addFeature("threads", "atomic operations",
+                    FeatureSet::Atomics)
+        .addFeature("mutable-globals", "mutable globals",
+                    FeatureSet::MutableGlobals)
+        .addFeature("nontrapping-float-to-int",
+                    "nontrapping float-to-int operations",
+                    FeatureSet::TruncSat)
+        .addFeature("simd", "SIMD operations and types",
+                    FeatureSet::SIMD)
+        .addFeature("bulk-memory", "bulk memory operations",
+                    FeatureSet::BulkMemory)
+        .add("--no-validation", "-n",
+             "Disables validation, assumes inputs are correct",
              Options::Arguments::Zero,
              [this](Options* o, const std::string& argument) {
                passOptions.validate = false;
              });
-        ;
   }
 
-  FeatureSet getFeatures() const {
+  ToolOptions &addFeatureOpt(const std::string& longName,
+                             const std::string& shortName,
+                             const std::string& description,
+                             Builder builder) {
+    add(longName, shortName, description, Options::Arguments::Zero,
+        [=](Options*, const std::string&) {
+          featureBuilders.push_back(builder);
+        });
+    return *this;
+  }
+
+  ToolOptions &addFeature(const std::string& name,
+                          const std::string& description,
+                          FeatureSet::Feature feature) {
+    (*this)
+        .addFeatureOpt(std::string("--enable-") + name, "",
+                       std::string("Enable ") + description,
+                       [=](FeatureSet& features) {
+                         features.set(feature, true);
+                       })
+        .addFeatureOpt(std::string("--disable-") + name, "",
+                       std::string("Disable ") + description,
+                       [=](FeatureSet& features) {
+                         features.set(feature, false);
+                       });
+    return *this;
+  }
+
+  FeatureSet& getFeatures(const Module& module) {
+    if (featuresInitialized) {
+      return passOptions.features;
+    }
+
+    bool wasmHasFeatures = module.readFeatures(wasmFeatures);
+    if (featureBuilders.size()) {
+      for (auto& builder : featureBuilders) {
+        builder(passOptions.features);
+      }
+      if (errorOnFeatureMismatch &&
+          !(wasmFeatures <= passOptions.features)) {
+        Fatal() << "module uses features not explicitly specified, "
+                << "use --detect-features to resolve";
+      }
+    } else if (wasmHasFeatures) {
+      passOptions.features = wasmFeatures;
+    }
+    featuresInitialized = true;
     return passOptions.features;
   }
+
+private:
+  std::vector<Builder> featureBuilders;
```

Comment:
I have to say that a vector of lambdas scares me, because in C++ there isn't any type checking to make sure they didn't close on something that went away.

I don't understand why we need the lambdas, though - instead we can write user-requested features to say `userEnabledFeatures` and `userDisabledFeatures`, and then use that later in the computation?